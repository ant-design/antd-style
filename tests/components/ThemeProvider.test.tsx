import { act, render, renderHook } from '@testing-library/react';
import { App, MappingAlgorithm, theme } from 'antd';
import { GetCustomToken, ThemeProvider, css, cx, useTheme, useThemeMode } from 'antd-style';
import { MessageInstance } from 'antd/es/message/interface';
import { NotificationInstance } from 'antd/es/notification/interface';
import { FC, PropsWithChildren } from 'react';
import { vi } from 'vitest';

interface TestDesignToken {
  customBrandColor: string;

  customColor?: string;

  a?: string;
}
interface TestDesignStylish {
  defaultText: string;
  x?: any;
}

declare module 'antd-style' {
  // eslint-disable-next-line @typescript-eslint/no-empty-interface
  interface CustomToken extends TestDesignToken {}

  // eslint-disable-next-line @typescript-eslint/no-empty-interface
  interface CustomStylish extends TestDesignStylish {}
}

describe('ThemeProvider', () => {
  describe('注入 antd 主题', () => {
    it('注入 token', () => {
      const Wrapper: FC<PropsWithChildren> = ({ children }) => (
        <ThemeProvider theme={{ token: { colorPrimary: '#000' } }}>{children}</ThemeProvider>
      );
      const { result } = renderHook(useTheme, { wrapper: Wrapper });

      expect(result.current.colorPrimary).toEqual('#000000');
    });

    it('仅注入算法', () => {
      const Wrapper: FC<PropsWithChildren> = ({ children }) => (
        <ThemeProvider theme={{ algorithm: theme.compactAlgorithm }}>{children}</ThemeProvider>
      );
      const { result } = renderHook(useTheme, { wrapper: Wrapper });

      expect(result.current.fontSize).toEqual(12);
    });

    it('注入暗色模式的自定义算法', () => {
      /**
       * 自定义主题算法
       * @param seedToken
       * @param mapToken
       */
      const customDarkAlgorithm: MappingAlgorithm = (seedToken, mapToken) => {
        const mergeToken = theme.darkAlgorithm(seedToken, mapToken);

        return {
          ...mergeToken,
          // Layout 颜色
          colorBgLayout: '#20252b',
          // 容器颜色
          colorBgContainer: '#282c34',
          // 悬浮类面板颜色
          colorBgElevated: '#32363e',
        };
      };

      const Wrapper: FC<PropsWithChildren> = ({ children }) => (
        <ThemeProvider
          appearance={'dark'}
          theme={(mode) =>
            mode === 'dark'
              ? { algorithm: [theme.compactAlgorithm, customDarkAlgorithm] }
              : undefined
          }
        >
          {children}
        </ThemeProvider>
      );
      const { result } = renderHook(useTheme, { wrapper: Wrapper });
      expect(result.current.colorPrimary).toEqual('#1668dc');
      expect(result.current.colorBgLayout).toEqual('#20252b');
      expect(result.current.fontSize).toEqual(12);
    });
  });

  describe('注入自定义主题', () => {
    it('自定义 Token', () => {
      const App = () => {
        const theme = useTheme();
        return <div style={{ color: theme.customBrandColor }}>{theme.customBrandColor}</div>;
      };

      const { container } = render(
        <ThemeProvider<TestDesignToken> customToken={{ customBrandColor: '#c956df' }}>
          <App />
        </ThemeProvider>,
      );

      expect(container).toMatchSnapshot();
      expect(container.firstChild).toHaveStyle({ color: '#c956df' });
    });

    it('自定义 Stylish', () => {
      const App = () => {
        const theme = useTheme();
        return <div className={cx(css(theme.stylish.defaultText))}>普通文本</div>;
      };

      const { container } = render(
        <ThemeProvider<any, TestDesignStylish>
          // @ts-ignore
          customStylish={({ css }) => ({
            defaultText: css`
              font-size: 14px;
              font-weight: 500;
              color: #333;
            `,
          })}
        >
          <App />
        </ThemeProvider>,
      );

      expect(container).toMatchSnapshot();
      expect(container.firstChild).toHaveStyle({ color: '#333', fontSize: 14, fontWeight: 500 });
    });

    it('注入自定义 Token 方法', () => {
      const customTokenFn: GetCustomToken<any> = ({ token, isDarkMode }) => ({
        customColor: isDarkMode ? '#000' : token.colorPrimary,
        customBrandColor: isDarkMode ? token.colorPrimary : '#FFF',
      });

      const Wrapper: FC<PropsWithChildren> = ({ children }) => (
        <ThemeProvider customToken={customTokenFn}>{children}</ThemeProvider>
      );

      const { result: light } = renderHook(useTheme, { wrapper: Wrapper });
      expect(light.current.customColor).toEqual('#1677ff');
      expect(light.current.customBrandColor).toEqual('#FFF');

      const Darker: FC<PropsWithChildren> = ({ children }) => (
        <ThemeProvider appearance={'dark'} customToken={customTokenFn}>
          {children}
        </ThemeProvider>
      );

      const { result: dark } = renderHook(useTheme, { wrapper: Darker });
      expect(dark.current.customColor).toEqual('#000');
      expect(dark.current.customBrandColor).toEqual('#1668dc');
    });

    it('注入自定义 stylish 方法', () => {
      const Wrapper: FC<PropsWithChildren> = ({ children }) => (
        <ThemeProvider customStylish={({ css }) => ({ x: css`` })}>{children}</ThemeProvider>
      );

      const { result } = renderHook(useTheme, { wrapper: Wrapper });
      expect(result.current.stylish.x!.styles).toEqual('');
    });

    it('嵌套 ThemeProvider 下的 自定义 Token', () => {
      const customTokenFn: GetCustomToken<any> = ({ token, isDarkMode }) => ({
        customColor: isDarkMode ? '#000' : token.colorPrimary,
        customBrandColor: isDarkMode ? token.colorPrimary : '#FFF',
      });

      const Wrapper: FC<PropsWithChildren> = ({ children }) => (
        <ThemeProvider customToken={customTokenFn}>
          <ThemeProvider customToken={{ a: 'red' }}>{children}</ThemeProvider>
        </ThemeProvider>
      );

      const { result: light } = renderHook(useTheme, { wrapper: Wrapper });
      expect(light.current.customColor).toEqual('#1677ff');
      expect(light.current.customBrandColor).toEqual('#FFF');
      expect(light.current.a).toEqual('red');
    });

    it('嵌套 ThemeProvider 算法', () => {
      const Wrapper: FC<PropsWithChildren> = ({ children }) => (
        <ThemeProvider
          theme={{
            algorithm: (designToken, derivativeToken) => ({
              ...designToken,
              ...derivativeToken!,
              colorBgLayout: '#213da2',
            }),
          }}
          customToken={{ a: 'red', customColor: 'world' }}
        >
          <ThemeProvider
            theme={{
              algorithm: (designToken, derivativeToken) => ({
                ...designToken,
                ...derivativeToken!,
                colorBgLayout: '#123ffa',
              }),
            }}
            customToken={{ a: 'ttt' }}
          >
            {children}
          </ThemeProvider>
        </ThemeProvider>
      );

      const { result: light } = renderHook(useTheme, { wrapper: Wrapper });
      expect(light.current.colorBgLayout).toEqual('#123ffa');
      expect(light.current.a).toEqual('ttt');
      expect(light.current.customColor).toEqual('world');
    });
  });

  describe('主题切换', () => {
    beforeEach(() => {
      Object.defineProperty(window, 'matchMedia', {
        writable: true,
        value: vi.fn().mockImplementation((query) => ({
          matches: true,
          media: query,
          onchange: null,

          addEventListener: vi.fn(),
          removeEventListener: vi.fn(),
          dispatchEvent: vi.fn(),
        })),
      });
    });

    it('自动模式', () => {
      const Wrapper: FC<PropsWithChildren> = ({ children }) => (
        <ThemeProvider themeMode={'auto'}>{children}</ThemeProvider>
      );

      const { result } = renderHook(useThemeMode, { wrapper: Wrapper });

      expect(result.current.themeMode).toEqual('auto');
    });
  });

  it('配合 App 实现局部作用域', async () => {
    const Demo: FC<{ id?: string }> = ({ id }) => {
      return (
        <div data-testid={id} style={{ padding: 16 }} className={'container'}>
          <a href="">节点样式</a>
        </div>
      );
    };

    const { container, findByTestId } = render(
      <>
        <Demo id={'without'} />
        <ThemeProvider>
          <App
            className={cx(css`
              .container {
                color: rgb(255, 0, 0);
              }
            `)}
          >
            <Demo id={'within'} />
          </App>
        </ThemeProvider>
        ,
      </>,
    );
    expect(container).toMatchSnapshot();

    const nodeWithout = await findByTestId('without');
    const nodeWithin = await findByTestId('within');

    expect(nodeWithin).toHaveStyle('color: rgb(255, 0, 0);');
    expect(nodeWithout).not.toHaveStyle('color: rgb(255, 0, 0);');
  });

  describe('静态实例对象', () => {
    it('获得静态实例对象', () => {
      let message = {} as MessageInstance;
      const Demo: FC = () => {
        return (
          <ThemeProvider getStaticInstance={(instances) => (message = instances.message)}>
            <div style={{ padding: 16 }} className={'container'}>
              <a href="">节点样式</a>
            </div>
          </ThemeProvider>
        );
      };

      render(<Demo />);

      expect(message.success).not.toBeUndefined();
    });
  });

  it('测试 prefix', () => {
    let message = {} as MessageInstance;
    let notification = {} as NotificationInstance;

    const Demo: FC = () => {
      return (
        <ThemeProvider
          prefixCls={'demo'}
          getStaticInstance={(instances) => {
            message = instances.message;
            notification = instances.notification;
          }}
          staticInstanceConfig={{
            message: {
              maxCount: 1,
              getContainer: () => document.getElementById('xxx')!,
            },
            notification: {
              getContainer: () => document.getElementById('xxx')!,
            },
          }}
        >
          <div id={'xxx'} style={{ padding: 16 }} className={'container'}>
            <a href="">节点样式</a>
          </div>
        </ThemeProvider>
      );
    };

    const { container } = render(<Demo />);

    act(() => {
      message.success('success');
    });

    expect(container.getElementsByClassName('demo-message')).toHaveLength(1);

    act(() => {
      notification.warning({ message: '123' });
    });

    expect(container.getElementsByClassName('demo-notification')).toHaveLength(1);
  });
});
