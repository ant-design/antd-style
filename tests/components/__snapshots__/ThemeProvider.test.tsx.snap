// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`ThemeProvider > 注入自定义主题 > 自定义 Stylish 1`] = `
.emotion-0 {
  font-size: 14px;
  font-weight: 500;
  color: #333;
}

<div>
  <div
    class="emotion-0"
  >
    普通文本
  </div>
</div>
`;

exports[`ThemeProvider > 注入自定义主题 > 自定义 Token 1`] = `
<div>
  <div
    style="color: rgb(201, 86, 223);"
  >
    #c956df
  </div>
</div>
`;

exports[`ThemeProvider > 配合 App 实现局部作用域 1`] = `
.emotion-0 .container {
  color: red;
}

<div>
  <div
    class="container"
    data-testid="without"
    style="padding: 16px;"
  >
    <a
      href=""
    >
      节点样式
    </a>
  </div>
  <div
    class="ant-app emotion-0"
  >
    <div
      class="container"
      data-testid="within"
      style="padding: 16px;"
    >
      <a
        href=""
      >
        节点样式
      </a>
    </div>
  </div>
  ,
</div>
`;
